crud.rb

```diff
@@ -1,69 +1,171 @@
+  def initialize
+    super
+    @users = DB[:users] #=> Sequel::Dataset
+  end
 
-# Displays a new user form (DONE)
-#-----------------------------------------------------------------
-get '/users/new' do
-  erb :"users/new"
-end
+  def user_signed_in?
+    !!session[:current_user_id]
+  end
 
-# Displays Home Page - Lists all the users in the database (DONE)
-#-----------------------------------------------------------------
-get '/users' do
-  @users.all #<---- is this necessary? still works without it.
-  erb :"users/index"
-end
+  def current_user
+    @current_user ||= @users.where(id: session[:current_user_id]).first
+  end
 
-# Displays specific users (DONE)
-#-----------------------------------------------------------------
-get '/users/:id' do
-  @user = @users.where(id: params[:id]).first
-  # @users = @users.find(:id => params[:id]) #<--- using this only displays first data in DB
-  erb :"users/show"
-end
+  def authorize_user(params=nil)
+    unless block_given?
+      raise "block must be provided"
+    end
 
-# Dispalys a user Edit Page (DONE)
-#-----------------------------------------------------------------
-get '/users/:id/edit' do
-  @user = @users.where(id: params[:id]).first
-  erb :"users/edit"
-end
+    if user_signed_in?
+      if current_user[:admin] == true
+        yield
+      elsif params.to_s == session[:current_user_id].to_s
+        yield
+      else
+        flash[:error] = "Unauthorized access"
+        redirect "/users/#{session[:current_user_id]}"
+      end
+    else  
+      flash[:error] = "Unauthorized access"
+      redirect '/homepage'
+    end
+  end
 
-# Creates new user and saves it to database (DONE)
-#-----------------------------------------------------------------
-post '/users' do
-  @users.insert(fname: params[:fname], lname: params[:lname], email: params[:email])
-  redirect '/users'
-end
+  # redirects '/' route to '/users' route (DONE)
+  #-----------------------------------------------------------------
+  get '/' do
+    redirect '/homepage'
+  end
 
-# Updates/Edits info of user (DONE)
-#-----------------------------------------------------------------
-patch '/users/:id' do
-  @user = @users.where(id: params[:id])
-  @user.update(fname: params[:fname], lname: params[:lname], email: params[:email])
-  redirect '/users'
-end
+  # Displays homepage for users to log in or sign up
+  #-----------------------------------------------------------------
+  get '/homepage' do
+    if user_signed_in?
+      flash[:error] = "Please log out to access homepage"
+      redirect "/users/#{session[:current_user_id]}"
+    else
+      erb :"homepage/index"
+    end
+  end
+
+  # Displays a new user form (DONE)
+  #-----------------------------------------------------------------
+  get '/users/new' do
+    erb :"users/new"
+  end
+
+  # Displays Home Page - Lists all the users in the database (DONE)
+  #-----------------------------------------------------------------
+  get '/users' do
```

Comment:
For this to work with a trailing slash, write it like:

`get '/users/?' do`

You can do the same for other paths like `'/homepage'`